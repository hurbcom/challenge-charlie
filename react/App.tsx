import React, { Suspense } from "react";
import ErrorBoundary from "./ErrorBoundary";
import { GlobalStyle } from "./styles/global";
import HeroImage from "@components/HeroImage";
import WeatherAccordeon from "@components/WeatherAccordeon";

const App: React.FC = () => {
    return (
        <Suspense fallback={<div>Suspending...</div>}>
            <ErrorBoundary>
                <GlobalStyle />
                <HeroImage
                    image={
                        "https://www.bing.com/th?id=OHR.KitsAspen_PT-BR8299899730_1920x1080.jpg&rf=LaDigue_1920x1080.jpg&pid=hp"
                    }
                >
                    <WeatherAccordeon />
                </HeroImage>
            </ErrorBoundary>
        </Suspense>
    );
};

export default App;
