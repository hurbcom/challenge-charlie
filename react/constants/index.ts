export const POSITIONS_TO_OPACITIES: Record<number, string> = {
    0: "--max-bg-opacity",
    1: "--medium-bg-opacity",
    2: "--min-bg-opacity",
};

export const SIZE = {
    mobileS: "320px",
    mobileM: "375px",
    mobileL: "425px",
    tablet: "768px",
    laptop: "1024px",
    laptopL: "1440px",
    desktop: "2560px",
};

export const DEVICES = {
    mobileS: `(min-width: ${SIZE.mobileS})`,
    mobileM: `(min-width: ${SIZE.mobileM})`,
    mobileL: `(min-width: ${SIZE.mobileL})`,
    tablet: `(min-width: ${SIZE.tablet})`,
    laptop: `(min-width: ${SIZE.laptop})`,
    laptopL: `(min-width: ${SIZE.laptopL})`,
    desktop: `(min-width: ${SIZE.desktop})`,
    desktopL: `(min-width: ${SIZE.desktop})`,
};
