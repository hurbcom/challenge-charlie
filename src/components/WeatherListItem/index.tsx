import styled, { css } from "styled-components";
import Icon from "components/Icon";

export type WeatherItemDataType = {
    day: string;
    temperature: number;
    weatherDescription: string;
    windSpeed: number;
    windDirection: string;
    pressure: number;
    humidity: number;
};

export type WeatherItemType = {
    compact?: boolean;
    data: WeatherItemDataType;
};

const isYellowContrastRange = (temperature: number) =>
    temperature > 14 && temperature < 23;

const getDayColor = (temperaure: number) => {
    const colors = [
        "#002e3c",
        "#003c4f",
        "#004d66",
        "#006382",
        "#006d8f",
        "#338aa5",
        "#549db4",
        "#8abccb",
        "#9BC2C2",
        "#ffe58a",
        "#fed954",
        "#fed233",
        "#fec700",
        "#e7b500",
        "#b48d00",
        "#8c6d00",
        "#6b5400",
        "#ff7f61",
        "#ff6642",
        "#ff4013",
        "#e83a11",
        "#b52d0d",
        "#8c230a",
    ];

    if (temperaure < 1) return colors[0];

    const temperatureColorIndex = Math.ceil(temperaure * 0.25) * 2 + 1;
    return temperatureColorIndex < colors.length
        ? colors[temperatureColorIndex]
        : colors[colors.length - 1];
};

const WheaterIcon = styled.div`
    display: flex;
    height: 155;
    width: 155;
    align-items: center;
    justify-content: center;

    & > * {
        font-size: 9rem;
        color: #ffffff;
    }
`;

const WheaterDetails = styled.ol<{ temperature: number }>`
    list-style-type: none;
    flex-grow: 1;
    padding: 16px 0;

    font-size: 12px;
    font-weight: 300;
    ${({ temperature }) =>
        isYellowContrastRange(temperature) &&
        css`
            color: #7b7508;
        `};

    & :nth-child(-n + 3) {
        font-size: 16px;
    }

    & :nth-child(-n + 2) {
        text-transform: uppercase;
    }

    & :nth-child(3) {
        margin-top: 16px;
        margin-bottom: 8px;
        text-transform: capitalize;
    }
`;

const WeatherListItemBase = styled.div<{ temperature: number }>`
    display: grid;
    grid-column-gap: 16px;
    grid-template-columns: repeat(2, 1fr);
    background-color: ${(props) => getDayColor(props.temperature)};
    opacity: 0.75;
    border-bottom: solid 1px rgba(0, 0, 0, 0.05);
`;

const WeatherListItem = ({
    compact = true,
    data,
    ...props
}: WeatherItemType) => {
    return (
        <WeatherListItemBase
            temperature={data?.temperature}
            role="presentation"
        >
            <WheaterIcon>{!compact ? <Icon name="sunny" /> : null}</WheaterIcon>
            <WheaterDetails temperature={data?.temperature}>
                <li>{data?.day}</li>
                <li>{Math.ceil(data?.temperature)}</li>
                {compact ? null : (
                    <>
                        <li>{data?.weatherDescription}</li>
                        <li>
                            Vento: {data?.windDirection} {data?.windSpeed}km/h
                        </li>
                        <li>Humidade: {data?.humidity}%</li>
                        <li>Press√£o: {data?.pressure}hPA</li>
                    </>
                )}
            </WheaterDetails>
        </WeatherListItemBase>
    );
};

export default WeatherListItem;
