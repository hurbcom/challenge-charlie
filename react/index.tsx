import React, { StrictMode } from "react";
import { hydrateRoot } from "react-dom/client";
import App from "./App";

const root = document.getElementById("react-app");

if (root) {
    hydrateRoot(
        root,
        <StrictMode>
            <App />
        </StrictMode>
    );
}
