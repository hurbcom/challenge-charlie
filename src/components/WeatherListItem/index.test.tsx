import { render, screen } from "@testing-library/react";
import WeatherListItem from ".";

const data = {
    day: "Hoje ",
    humidity: 70,
    pressure: 1015,
    temperature: 24.28,
    windDirection: "SW",
    windSpeed: 3.39,
    weatherDescription: "chuva leve",
};

describe("<WeatherListItem />", () => {
    test("render compact weather item", () => {
        const { asFragment } = render(<WeatherListItem data={data} />);
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.queryByText(/chuva leve/)).not.toBeInTheDocument();
    });
    test("render full weather item", () => {
        const { asFragment } = render(
            <WeatherListItem data={data} compact={false} />
        );
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.getByText(/chuva leve/)).toBeInTheDocument();
    });

    test("render weather item with zero temperature", () => {
        const { asFragment } = render(
            <WeatherListItem
                data={{
                    ...data,
                    temperature: 0,
                }}
            />
        );
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.queryByText(/chuva leve/)).not.toBeInTheDocument();
        expect(screen.getByRole("presentation")).toHaveStyleRule(
            "background-color",
            "#002e3c"
        );
    });

    test("render weather item with yellow range temperature deg", () => {
        const { asFragment } = render(
            <WeatherListItem
                data={{
                    ...data,
                    temperature: 20,
                }}
            />
        );
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.queryByText(/chuva leve/)).not.toBeInTheDocument();
        expect(screen.getByRole("presentation")).toHaveStyleRule(
            "background-color",
            "#fed233"
        );
    });

    test("render weather item with 44 temperature render last red color", () => {
        const { asFragment } = render(
            <WeatherListItem
                data={{
                    ...data,
                    temperature: 44,
                }}
            />
        );
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.queryByText(/chuva leve/)).not.toBeInTheDocument();
        expect(screen.getByRole("presentation")).toHaveStyleRule(
            "background-color",
            "#8c230a"
        );
    });

    test("render weather item with 40 temperature render second-last red color", () => {
        const { asFragment } = render(
            <WeatherListItem
                data={{
                    ...data,
                    temperature: 40,
                }}
            />
        );
        expect(asFragment).toMatchSnapshot();

        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByText(/Hoje/)).toBeInTheDocument();
        expect(screen.queryByText(/chuva leve/)).not.toBeInTheDocument();
        expect(screen.getByRole("presentation")).toHaveStyleRule(
            "background-color",
            "#b52d0d"
        );
    });
});
