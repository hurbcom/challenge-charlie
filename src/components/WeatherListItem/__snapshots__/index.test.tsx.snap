// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherListItem /> render compact weather item 1`] = `[Function]`;

exports[`<WeatherListItem /> render full weather item 1`] = `[Function]`;

exports[`<WeatherListItem /> render weather item with 40 temperature render second-last red color 1`] = `[Function]`;

exports[`<WeatherListItem /> render weather item with 44 temperature render last red color 1`] = `[Function]`;

exports[`<WeatherListItem /> render weather item with yellow range temperature deg 1`] = `[Function]`;

exports[`<WeatherListItem /> render weather item with zero temperature 1`] = `[Function]`;
